lib/llm/src/protocols/openai/chat_completions/jail.rs

```diff
@@ -745,6 +764,44 @@ impl JailedStream {
         }
         false
     }
+
+    /// Post-processor that sets finish_reason to ToolCalls when tool calls were emitted
+    /// This should be called after apply() to fix the finish_reason for tool call chunks
+    pub fn fix_finish_reason<S>(
+        input_stream: S,
+    ) -> impl Stream<Item = Annotated<NvCreateChatCompletionStreamResponse>> + Send
+    where
+        S: Stream<Item = Annotated<NvCreateChatCompletionStreamResponse>> + Send + 'static,
+    {
+        stream! {
+            tokio::pin!(input_stream);
```

Comment:
Food for thought for a future PR - why do we need to process the whole stream again to fix the finish reason? apply() is already iterating through the stream, right? Can we fix the finish reason in one shot while going through the stream the first time?

Reply:
Please make a ticket for follow up so we don't forget, but not an immediate priority over the multimodal disagg migration